<a href="https://colab.research.google.com/github/CrysisDavid/ProgramacionLineal/blob/main/Taller2_SymPy_SciPy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Corporación Universitaria Remington
## Ingeniería de Sistemas
### Prof. Diego Fernando Marin -  ```diego.marin@uniremington.edu.co```

## Presentado por:

- Nombre: *Nombre completo del estudiante*
- Correo: `xxxxxxxx@miremington.edu.co`

# SymPy

En los lenguajes de programación es común encontrar funciones matemáticas, pero están orientadas a calcular el resultado numérico de la función, por lo que son útiles para el análisis matemático de las mismas.

Con [SymPy](https://www.sympy.org/), en la computación (o matemática) simbólica, las funciones mantienen la expresión sin calcular el resultado numérico. Estas expresiones pueden ser utilizadas y combinadas con otras expresiones, pero además es posible aplicar operaciones sobre las expresiones y en cada caso obtener otra expresión como resultado.

Si es necesario, también se puede obtener el resultado numérico reemplazando los simbolos con valores.

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import math

In [ ]:
# la forma convencional de importar la librería SymPy
import sympy as sy

sy.__version__

'1.12'

## diferencia cálculo númerico vs. simbólico

In [ ]:
# cálculo numérico - racionales


In [ ]:
# cálculo simbólico - racionales


In [ ]:
# cálculo numérico - raiz cuadrada


In [ ]:
# cálculo simbólico - raiz cuadrada


## Símbolos

In [ ]:
# creación de símbolos


In [ ]:
type(x)

## Constantes

## Ecuaciones

## Factorización & Expansión

## Integración & Derivada

## Sistemas de ecuaciones

# SciPy

[SciPy](https://scipy.org/) es una colección de algoritmos matemáticos y las funciones que permiten su uso, construidas sobre ```NumPy``` y que proporcionan las clases y métodos para procesar datos científicos de la misma forma que MATLAB, [Octave](https://octave.org/), [SciLab](https://www.scilab.org/) y similares.

In [ ]:
# la forma convencional de importar la librería SciPy
import scipy as sp
from scipy import stats
from scipy import optimize

sp.__version__

## Estadística

In [ ]:
# genere 500 enteros aleatorios entre 0 y 1000
xs = np.random.randint(0, 1000, (500))
xs[:50] # los primeros 50

In [ ]:
# mínimo, máximo, sumatoria
xs.min(), xs.max(), xs.sum()

In [ ]:
# media, varianza y desviación estandar
xs.mean(), xs.var(), xs.std()

In [ ]:
# media geométrica: raíz n-ésima del producto de todos los números
stats.gmean(xs)

In [ ]:
# media armónica: inverso de la media aritmética
stats.hmean(xs)

In [ ]:
# skewness: grado de asimetría de la distribución
stats.skew(xs)

In [ ]:
# kurtosis: grado de concentración de los valores
stats.kurtosis(xs)

In [ ]:
stats.describe(xs)

## Encontrar el mínimo de una función

In [ ]:
def f(x):
    return x**2 + 10*np.sin(x)

In [ ]:
# usando el método Nelder Mead, por defecto


In [ ]:
# usando el método BFGS, L=memoria limitada, B=variación Nocedal-Byrd


La cadena `'ro'` representa una marca roja en forma circular. Otras opciones para color son `rgbycmk`, y para las marcas tenemos `o.sd*^`.

# Ejercicios

In [ ]:
x, y, z = sy.symbols("x y z")

1.1 Simplifique la expresión de la siguiente ecuación:

$${ x^3 + x^2 - x - 1 } \over { x^2 + 2x + 1 }$$

1.2 Genere los términos de la siguiente ecuación:

$$3x (4 - x) + (2x - 5)^2$$

1.3 Factorice la siguiente ecuación:

$$ 6x^4y + 15x^3y − 9x^2y $$

1.4 Encuentre la solución para la siguiente ecuación:
$$3x^2 − 8 = 100$$

1.5 Encuentre la solución para la siguiente ecuación:
$$2x^3 + 16 = 0$$

1.6 Encuentre la solución para la siguiente ecuación:

$$\left (  2x + {y \over 5} \right ) ^ 3 = 0$$

1.7 Verifique que la siguiente igualdad es cierta:

$$ {d \over dx} tan(x) = \lim_{y \,\to\, 0} {{tan(x+y)-tan(x)} \over y } $$

1.8. Calcule el valor de la siguiente integral:

$$\int_{-\infty}^\infty {\sin x^2  dx}$$


1.9 Encuentre la solución para el siguiente sistema de ecuaciones:

$$ x + 5y = 2$$
$$-3x + 6y = 15$$

1.10 Encuentre la solución para el siguiente sistema de ecuaciones:

$$3x + {y \over2} - 5 = 0$$
$$x^2 - 2y + 7 = 0$$

1.11 Encuentre y grafique el valor mínimo (vertice) de la siguiente función cuadrática:

$$x^2 + 2x + 3$$

**Nota**:
1. Saque una copia con la opción *Archivo* → *Guardar una copia en Drive*.
2. Renombre el *cuaderno* a: `solucion_Taller2_SymPy_SciPyNombreCompleto.ipynb`.
3. Ejecute todas las celdas para revisar los temas del taller.
4. Solucione todos los puntos del taller.
5. Comparta la solución y envié el link al profesor, antes de la fecha limite (no se aceptan fuera de tiempo).
6. No copie! la copia tiene nota **1.0**.